CARDS-2535: Add ability to set default values

veronikaslc opened the pull request:
[CARDS-2535](https://phenotips.atlassian.net/browse/CARDS-2535): Add ability to set default values

[CARDS-2535]: https://phenotips.atlassian.net/browse/CARDS-2535?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

veronikaslc committed:
CARDS-2535: Add ability to set default values

veronikaslc committed:
CARDS-2535: Add ability to set default values

veronikaslc committed:
CARDS-2535: Add ability to set default values
Let display components use defaults

veronikaslc committed:
CARDS-2535: Add ability to set default values
Let display components use defaults

veronikaslc committed:
CARDS-2535: Add ability to set default values
Parse some default values

veronikaslc committed:
CARDS-2535: Add ability to set default values
Parse some default values

veronikaslc commented:
This is a draft PR to see what else is needed for parsing or refactoring.

veronikaslc committed:
[misc] Better codestyling for component props

marta- committed:
CARDS-2535: Apply default values in question matrices

Default values could be set on a question matrix and its entries in the
questionnaire editor, and were stored on the cards:Section and
cards:QuestionMatrixEntry nodes, but the matrix never used them when
rendering a form, so they had no effect. This wires them into QuestionMatrix.

marta- committed:
CARDS-2535: Pre-select the default value for multiple choice questions

The default value was placed in the input (when available) even if it
matched one of the predefined options.

marta- committed:
CARDS-2535: Parse the default value for time questions

A time question's default value was assigned to the selected time as the raw
string read from the question definition, but the rest of the component expects
a Luxon DateTime.

marta- committed:
CARDS-2535: Order the default value field after the display mode in the editor

Since the order of fields in the json files dictates the order in which they
are presented to the user in the EditDialog, ensure the defaultValue is after
dataMode / answerOptions wherever possible, for a better flow.

marta- committed:
CARDS-2535: Support comma-separated default values for multivalued questions

Allow the defaultValue of multivalued questions to be a comma-separated list.
The list is split on commas, trimmed, de-duplicated, and capped at maxAnswers.
Single-valued questions keep their existing behaviour.

marta- committed:
CARDS-2535: Support multiple and default values for number questions

A number question transformed its default value with Number() before using it,
which discarded a comma-separated list as not-a-number. As a result a
multivalued number question (a multi-select list of numeric options) could not
be given more than one default, and a number range could not be given default
bounds at all.

Parse the default into a list of numbers instead, dropping any entry that is not
a valid number.

marta- committed:
CARDS-2535: Apply the default value for date questions

A date question parsed and validated its configured default value but never
used it, so the default was ignored.

marta- committed:
CARDS-2535: Match an option default value by value or label

When resolving a default value against a question's predefined answer options,
match by either the internal value or the displayed label. An option is already
considered selected when rendered if its label or value matches, so matching
only by value was inconsistent:
- in a question with answer options and free-text entry, a default given as an
  option's label was kept as a separate custom value while that same option also
  rendered as selected, so it appeared twice;
- in a matrix, a default given as an option's label matched nothing and was
  silently discarded.

marta- committed:
CARDS-2535: Parse the date default using the question's date format

The date default was parsed with DateTime.fromISO, but the editor hint tells
users to enter it in the question's configured date format.

marta- committed:
CARDS-2535: De-duplicate option defaults that resolve to the same value

A multivalued default is de-duplicated on its raw comma-separated tokens, but
two distinct tokens can resolve to the same option when one is the option's
label and the other its value (e.g. "Yes, 1" for an option labelled "Yes" with
value "1"). That pre-selected the same option twice and could save a duplicate.

De-duplicate the resolved [label, value] selections by value, before applying
the maxAnswers cap, in both MultipleChoice and QuestionMatrix.

marta- committed:
CARDS-2535: Fix md syntax and improve wording of the defaultValue hint in Question-hints.json

marta- committed:
CARDS-2535: Concatenate question-type hints onto shared field hints

When a question type contributes a field hint whose field already has a hint in
the shared Question-hints.json, append the contribution to the existing hint
instead of replacing it.

The goal is to allow optional question types such as SelectableArea to contribute
their own hinfo to the defaultValue hint.

marta- committed:
CARDS-2535: Move the selectable area and vocabulary default value notes to their modules

The defaultValue hint documented every question type in one block in the shared
Question-hints.json, including selectableArea and vocabulary, which CARDS-2910
moved into their own modules.

Move those two notes next to their components, in SelectableAreaQuestion-editor-
hints.json and a new VocabularyQuestion-editor-hints.json, exposing the latter as
a questionEditorHints static. They are appended back onto the shared defaultValue
description at load time via the new mergeHints helper in QuestionModelManager.jsx.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/BooleanQuestion.jsx:
```diff
@@ -74,6 +74,7 @@ function BooleanQuestion(props) {
         valueType="Long" /* Notably not "Boolean", since we need it to be stored as a long in the backend */
         maxAnswers={1}
         defaults={options}
+        defaultValue={props.questionDefinition.defaultValue}
```
Doesn't work for booleans.

marta- replied:
Addressed in b295e77.  Also added a test questionnaire DefaultValueTest (commit 637445a) and expanded the default value support in BooleanQuestion to accept "true"/"false", "yes"/"no"/"unknown", yesLabel, etc in addition to the numeric values (in commit ef5524a).

marta- committed:
CARDS-2535: Accept common spellings for boolean default values

A boolean question only pre-selected its default when the value was the stored
1, 0 or -1 (or an exact option label).

Normalize the boolean default to the stored 1/0/-1 value, accepting 1/0/-1,
true/false, yes/no and the configured Yes/No/Unknown labels in any case.

marta- committed:
CARDS-2535: Add a test questionnaire

marta- committed:
CARDS-2535: Apply option/matrix defaults when the answer has no value yet

MultipleChoice and QuestionMatrix only pre-selected default values when the
question had no answer node at all. On a real form the answer nodes are
pre-created without a value, so the defaults were skipped and only showed in the
questionnaire preview (which has no answers).

Apply the default whenever there is no saved value yet: in MultipleChoice when
existingAnswer is absent or its value is undefined, and in QuestionMatrix per
row for any row not already populated from an existing answer.

sdumitriu reviewed (changes requested):
Problem: In the `Text, list + input, custom (non-option) default`, the custom default (purple) is filled in the input box, but it is not selected, which means that saving will leave the question unanswered.

Unfixable problem: `Text, list, multiple answers`, unselect all the values, save -> answer is missing (expected). Edit again, the defaults will be selected again, which could lead to unwanted answers. Without doing a big change, this is impossible to fix, and I think it's fine for now.

marta- committed:
CARDS-2535: Select a custom default value, not just fill it in the input

A single-answer question's default that matched no predefined option was filled
into the free-text (ghost) input but never added to the selection, so its radio
stayed unchecked and the answer saved as empty.
